### PR TITLE
Add replaceOnChange to schema

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Computes an appropriate `replaceOnChange` value for resources at codegen time. 

Check the property list for `replaceOnChange`. If a property is a `Resource`, recursively check its properties. 

Fixes https://github.com/pulumi/pulumi/issues/7798

This now allows 
```json
{
  "version": "0.0.1",
  "name": "example",
  "resources": {
    "example::Other": {
      "properties": {
        "field": {
          "$ref": "#/types/example::Dog"
        }
      }
    },
    "example::Dog": {
      "properties": {
        "bone": {
          "type": "string",
          "replaceOnChanges": true
        }
      }
    },
    "example::Cat": {
      "properties": {
        "name": {
          "type": "string",
          "replaceOnChanges": true
        },
        "other": {
          "$ref": "#/types/example::Other"
        }
      }
    }
  },
  "language": {
    "csharp": {},
    "go": {
      "generateResourceContainerTypes": true
    },
    "nodejs": {},
    "python": {}
  }
}
```
when defining a resource in a schema. I included the `${resurce}.ts` files generated.

<details><summary>cat.ts</summary>
<p>

```typescript
// *** WARNING: this file was generated by pulumictl. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***

import * as pulumi from "@pulumi/pulumi";
import * as utilities from "./utilities";

import {Other} from "./index";

export class Cat extends pulumi.CustomResource {
    /**
     * Get an existing Cat resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    public static get(name: string, id: pulumi.Input<pulumi.ID>, opts?: pulumi.CustomResourceOptions): Cat {
        return new Cat(name, undefined as any, { ...opts, id: id });
    }

    /** @internal */
    public static readonly __pulumiType = 'example::Cat';

    /**
     * Returns true if the given object is an instance of Cat.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    public static isInstance(obj: any): obj is Cat {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Cat.__pulumiType;
    }

    public /*out*/ readonly name!: pulumi.Output<string | undefined>;
    public /*out*/ readonly other!: pulumi.Output<Other | undefined>;

    /**
     * Create a Cat resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name: string, args?: CatArgs, opts?: pulumi.CustomResourceOptions) {
        let inputs: pulumi.Inputs = {};
        opts = opts || {};
        if (!opts.id) {
            inputs["name"] = undefined /*out*/;
            inputs["other"] = undefined /*out*/;
        } else {
            inputs["name"] = undefined /*out*/;
            inputs["other"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion()});
        }
        const replaceOnChanges = { replaceOnChanges: ["name", "other.field.bone"] };
        opts = pulumi.mergeOptions(opts, replaceOnChanges);
        super(Cat.__pulumiType, name, inputs, opts);
    }
}

/**
 * The set of arguments for constructing a Cat resource.
 */
export interface CatArgs {
}
```

</p>
</details>

<details><summary>dog.ts</summary>
<p>

```typescript
// *** WARNING: this file was generated by pulumictl. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***

import * as pulumi from "@pulumi/pulumi";
import * as utilities from "./utilities";

export class Dog extends pulumi.CustomResource {
    /**
     * Get an existing Dog resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    public static get(name: string, id: pulumi.Input<pulumi.ID>, opts?: pulumi.CustomResourceOptions): Dog {
        return new Dog(name, undefined as any, { ...opts, id: id });
    }

    /** @internal */
    public static readonly __pulumiType = 'example::Dog';

    /**
     * Returns true if the given object is an instance of Dog.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    public static isInstance(obj: any): obj is Dog {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Dog.__pulumiType;
    }

    public /*out*/ readonly bone!: pulumi.Output<string | undefined>;

    /**
     * Create a Dog resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name: string, args?: DogArgs, opts?: pulumi.CustomResourceOptions) {
        let inputs: pulumi.Inputs = {};
        opts = opts || {};
        if (!opts.id) {
            inputs["bone"] = undefined /*out*/;
        } else {
            inputs["bone"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion()});
        }
        const replaceOnChanges = { replaceOnChanges: ["bone"] };
        opts = pulumi.mergeOptions(opts, replaceOnChanges);
        super(Dog.__pulumiType, name, inputs, opts);
    }
}

/**
 * The set of arguments for constructing a Dog resource.
 */
export interface DogArgs {
}

```

</p>
</details>

<details><summary>other</summary>
<p>

```typescript
// *** WARNING: this file was generated by pulumictl. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***

import * as pulumi from "@pulumi/pulumi";
import * as utilities from "./utilities";

import {Dog} from "./index";

export class Other extends pulumi.CustomResource {
    /**
     * Get an existing Other resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    public static get(name: string, id: pulumi.Input<pulumi.ID>, opts?: pulumi.CustomResourceOptions): Other {
        return new Other(name, undefined as any, { ...opts, id: id });
    }

    /** @internal */
    public static readonly __pulumiType = 'example::Other';

    /**
     * Returns true if the given object is an instance of Other.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    public static isInstance(obj: any): obj is Other {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Other.__pulumiType;
    }

    public /*out*/ readonly field!: pulumi.Output<Dog | undefined>;

    /**
     * Create a Other resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name: string, args?: OtherArgs, opts?: pulumi.CustomResourceOptions) {
        let inputs: pulumi.Inputs = {};
        opts = opts || {};
        if (!opts.id) {
            inputs["field"] = undefined /*out*/;
        } else {
            inputs["field"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion()});
        }
        const replaceOnChanges = { replaceOnChanges: ["field.bone"] };
        opts = pulumi.mergeOptions(opts, replaceOnChanges);
        super(Other.__pulumiType, name, inputs, opts);
    }
}

/**
 * The set of arguments for constructing a Other resource.
 */
export interface OtherArgs {
}
```

</p>
</details>

Note that `dog` has `replaceOnChanges = ["bone"]`, `other` has `replaceOnChanges = ["field.bone"]`, and `cat` has `replaceOnChanges = ["name", "other.field.bone"]`.

## Checklist
- [ ] Update the [schema docs](https://www.pulumi.com/docs/guides/pulumi-packages/schema/#pulumi-package-schema)

- [x] Tests
    + [x] Test for `nodejs`
    + [x] Test for `python`
    + [x] Test for `dotnet`

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

### CODE REVIEW

File: pkg/codegen/schema/schema.go
Comment (justinvp) [inline pkg/codegen/schema/schema.go:403]: @lblackstone, should we be looking at the resource's output properties (`Properties`), input properties (`InputProperties`), or both?

File: pkg/codegen/schema/schema.go
Comment (justinvp) [inline pkg/codegen/schema/schema.go:406]: I don't think this is quite right. If a resource has a property that's typed as a `*ResourceType`, I don't think we want to drill into its properties. @lblackstone, correct me if I am wrong.

I think we do want to drill into properties typed as `*ObjectType` (looking at its `.Properties` deeply).

And, in addition to handling `*OptionalType`, I suspect we also need to handle `*UnionType`.

@lblackstone, what about `*MapType` and `*ArrayType`? Is there a syntax that `replaceOnChanges` supports for those?

**Review by justinvp — COMMENTED** [review]
Some questions/comments.

Also, looking at your example schema:

```json
{
  "version": "0.0.1",
  "name": "example",
  "resources": {
    "example::Other": {
      "properties": {
        "field": {
          "$ref": "#/types/example::Dog"
        }
      }
    },
    "example::Dog": {
      "properties": {
        "bone": {
          "type": "string",
          "replaceOnChanges": true
        }
      }
    },
    "example::Cat": {
      "properties": {
        "name": {
          "type": "string",
          "replaceOnChanges": true
        },
        "other": {
          "$ref": "#/types/example::Other"
        }
      }
    }
  },
  "language": {
    "csharp": {},
    "go": {
      "generateResourceContainerTypes": true
    },
    "nodejs": {},
    "python": {}
  }
}
```

Is this schema valid? The `"$ref": "#/types/example::Dog"` is referencing a type with token of `example::Dog`, but I don't see that type under `{"types":{"example::Dog":{}}}`, I only see another resource with a token of `example::Dog` under `{"resources":{"example::Dog":{}}}`, which would be referenced via `"$ref": "#/resources/example::Dog"`.

File: pkg/codegen/schema/schema.go
Comment (lblackstone) [reply pkg/codegen/schema/schema.go:403]: We should check both inputs and outputs.

File: pkg/codegen/schema/schema.go
Comment (lblackstone) [reply pkg/codegen/schema/schema.go:406]: > @lblackstone, what about `*MapType` and `*ArrayType`? Is there a syntax that `replaceOnChanges` supports for those?

Yes, those can be specified like this:
- Map: `['foo.bar', 'baz.*']`
- Array: `['foo[0]', 'bar[*]']`

Comment (iwahbe) [reply pkg/codegen/schema/schema.go:406]: @lblackstone, Is the idea that you can do something like this:
```json
"bar": {
  "type": "array",
  "items": {
    "type": "number",
    "replaceOnChanges": true,
  }
}
```
to generate `'bar[*]'`. 
```json
"bar": {
  "type": "array",
  "replaceOnChanges": true,
  "items": {
    "type": "number"
  }
}
```
already generates `'bar'`. 
And the same thing for maps?

Comment (lblackstone) [reply pkg/codegen/schema/schema.go:406]: Your example looks correct to me.

Also note that specifying the parent implicitly causes a change if any of the dependents change. For example,

`['']` or `['*']` will always cause a replacement since it specifies the root of the resource.

**Comment by iwahbe** [comment]
@justinvp and @lblackstone What is the desired behavior for the following object:
```json
"types": {
    "example::toy": {
      "type": "object",
      "properties": {
        "wear": {
          "type": "number",
          "replaceOnChanges": true
        },
        "pair": {
          "$ref": "#/types/example::toy",
        }
      }
    }
  },
```
Given that it's defined recursively, a "correct" result would be `[toy.wear, toy.pair.wear, toy.pair.pair.wear, ...]`, which is both infinite and impossible. I can detect the recursion, but I don't know what to spit out afterwards. All references are optional, so this doesn't mean that the program is invalid in practice.

**Comment by lblackstone** [comment]
That type of recursion is fairly rare; I think it's generally for stuff like schema definitions. For recursive cases, I think it would be ok to ignore nested properties. Concretely, I would generate `[toy.wear]` for your example. If we discover that this strategy misses important cases later, we can always revisit, but I think that should be good enough for now.

File: pkg/codegen/schema/schema.go
Comment (iwahbe) [reply pkg/codegen/schema/schema.go:420]: The comment on `Properties` says it should be a superset of `InputProperties`, so I only check `Properties`. Is this wrong?

File: pkg/codegen/schema/schema.go
Comment (iwahbe) [reply pkg/codegen/schema/schema.go]: This should be resolved now.

File: pkg/codegen/schema/schema.go
Comment (justinvp) [reply pkg/codegen/schema/schema.go:420]: > The comment on `Properties` says it should be a superset of `InputProperties`, so I only check `Properties`

That's generally the case, yes. I guess it depends on where we want the user to be able to specify these in the schema. And if we do end up looking at both, what happens if there's a conflict? @lblackstone, I'll defer to you on this.

File: pkg/codegen/internal/test/testdata/replace-on-change/schema.json
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/replace-on-change/schema.json:111]: Is this supposed to be `replaceOnChanges` plural?

**Review by justinvp — APPROVED** [review]
A couple more questions, but otherwise LGTM. Nice work. It would be good for @lblackstone, to take one last look before merging to ensure it looks good for what he needs for the Kubernetes provider.

**Comment by iwahbe** [comment]
I clicked merge and then saw the comments. The rest of this conversation can take place on https://github.com/pulumi/pulumi/pull/7921, which is the fixup for this merge.

File: pkg/codegen/schema/schema.go
Comment (lblackstone) [reply pkg/codegen/schema/schema.go:420]: I think it should be fine to only check `Properties`.

**Pull request merged**
